SU2_CFD/src/solvers/CSolver.cpp

```diff
@@ -2065,79 +2066,81 @@ void CSolver::AdaptCFLNumber(CGeometry **geometry,
-        New_Func += solverTurb->GetRes_RMS(iVar);
-      }
-    }
+      /* Sum the RMS residuals for all equations. */
 
-    /* Compute the difference in the nonlinear residuals between the
-     current and previous iterations. */
+      New_Func = 0.0;
+      for (unsigned short iVar = 0; iVar < solverFlow->GetnVar(); iVar++) {
+        New_Func += log10(solverFlow->GetRes_RMS(iVar));
+      }
+      if ((iMesh == MESH_0) && solverTurb) {
+        for (unsigned short iVar = 0; iVar < solverTurb->GetnVar(); iVar++) {
+          New_Func += log10(solverTurb->GetRes_RMS(iVar));
+        }
+      }
 
-    NonLinRes_Func = (New_Func - Old_Func);
-    NonLinRes_Series[NonLinRes_Counter] = NonLinRes_Func;
+      /* Compute the difference in the nonlinear residuals between the
+       current and previous iterations, taking care with very low initial
+       residuals (due to initialization). */
 
-    /* Increment the counter, if we hit the max size, then start over. */
+      if ((config->GetInnerIter() == 1) && (New_Func - Old_Func > 10)) {
+        Old_Func = New_Func;
+      }
+      NonLinRes_Series[NonLinRes_Counter] = New_Func - Old_Func;
 
-    NonLinRes_Counter++;
-    if (NonLinRes_Counter == Res_Count) NonLinRes_Counter = 0;
+      /* Increment the counter, if we hit the max size, then start over. */
 
-    /* Sum the total change in nonlinear residuals over the previous
-     set of all stored iterations. */
+      NonLinRes_Counter++;
+      if (NonLinRes_Counter == Res_Count) NonLinRes_Counter = 0;
 
-    NonLinRes_Value = New_Func;
-    if (config->GetTimeIter() >= Res_Count) {
-      NonLinRes_Value = 0.0;
-      for (unsigned short iCounter = 0; iCounter < Res_Count; iCounter++)
-        NonLinRes_Value += NonLinRes_Series[iCounter];
-    }
+      /* Detect flip-flop convergence to reduce CFL and large increases
+       to reset to minimum value, in that case clear the history. */
 
-    /* If the sum is larger than a small fraction of the current nonlinear
-     residual, then we are not decreasing the nonlinear residual at a high
-     rate. In this situation, we force a reduction of the CFL in all cells.
-     Reset the array so that we delay the next decrease for some iterations. */
+      if (config->GetInnerIter() >= Res_Count) {
+        unsigned long signChanges = 0;
+        su2double totalChange = 0.0;
+        auto prev = NonLinRes_Series.front();
+        for (auto val : NonLinRes_Series) {
+          totalChange += val;
+          signChanges += (prev > 0) ^ (val > 0);
+          prev = val;
+        }
+        reduceCFL |= (signChanges > Res_Count/4) && (totalChange > -0.5);
 
-    if (fabs(NonLinRes_Value) < 0.1*New_Func) {
-      NonLinRes_Counter = 0;
-      for (unsigned short iCounter = 0; iCounter < Res_Count; iCounter++)
-        NonLinRes_Series[iCounter] = New_Func;
+        if (totalChange > 2.0) { // orders of magnitude
+          resetCFL = true;
```

Comment:
And this is some simple logic to detect oscillations and possible divergence.

Reply:
Just to make sure I am following this correctly, this resets if the Non linear residual increases by 2 orders of magnitude? And it reduces the CFL if the non-linear residual oscillates 25% of the time AND there is less than a 0.5 order of magnitude reduction.

This seems to work as designed. Also I appreciate the flex with the XOR operator. Don't see that very often.

Reply:
Correct, if the residuals are going down it does not care if the history does not look pretty.
Bigger flex is to compute the total change by adding consecutive deltas, which still works when it wraps around to overwrite the oldest values.